core/src/main/scala/com/avast/clients/rabbitmq/publisher/PublishConfirmsRabbitMQProducer.scala

```diff
@@ -0,0 +1,103 @@
+import com.avast.metrics.scalaeffectapi.Monitor
+import com.rabbitmq.client.ConfirmListener
+
+class PublishConfirmsRabbitMQProducer[F[_], A: ProductConverter](name: String,
+                                                                 exchangeName: String,
+                                                                 channel: ServerChannel,
+                                                                 defaultProperties: MessageProperties,
+                                                                 sentMessages: SentMessages[F],
+                                                                 sendAttempts: Int,
+                                                                 reportUnroutable: Boolean,
+                                                                 sizeLimitBytes: Option[Int],
+                                                                 blocker: Blocker,
+                                                                 logger: ImplicitContextLogger[F],
+                                                                 monitor: Monitor[F])(implicit F: ConcurrentEffect[F], cs: ContextShift[F])
+    extends BaseRabbitMQProducer[F, A](name,
+                                       exchangeName,
+                                       channel,
+                                       defaultProperties,
+                                       reportUnroutable,
+                                       sizeLimitBytes,
+                                       blocker,
+                                       logger,
+                                       monitor) {
+
+  channel.confirmSelect()
+  channel.addConfirmListener(DefaultConfirmListener)
+
+  private val acked = monitor.meter("acked")
+  private val nacked = monitor.meter("nacked")
+
+  override def sendMessage(routingKey: String, body: Bytes, properties: MessageProperties)(implicit correlationId: CorrelationId): F[Unit] =
+    sendWithAck(routingKey, body, properties, 1)
+
+  private def sendWithAck(routingKey: String, body: Bytes, properties: MessageProperties, attemptCount: Int)(
+      implicit correlationId: CorrelationId): F[Unit] = {
+
+    if (attemptCount > sendAttempts) {
+      F.raiseError(MaxAttempts("Exhausted max number of attempts"))
+    } else {
+      val messageId = channel.getNextPublishSeqNo
+      for {
+        defer <- Deferred.apply[F, Either[Throwable, Unit]]
+        _ <- sentMessages.update(_ + (messageId -> defer))
+        _ <- basicSend(routingKey, body, properties)
+        result <- defer.get
+        _ <- result match {
+          case Left(err) =>
+            val sendResult = if (sendAttempts > 1) {
+              clearProcessedMessage(messageId) >> sendWithAck(routingKey, body, properties, attemptCount + 1)
+            } else {
+              F.raiseError(NotAcknowledgedPublish(s"Broker did not acknowledge publish of message $messageId", err))
+            }
```

Comment:
Here you "catch" an arbitrary `Throwable` and say "it was not acked" then.. Maybe change the `Either` accordingly?